src/prof.c

```diff
@@ -106,6 +108,8 @@ static uint64_t		prof_dump_mseq;
 static uint64_t		prof_dump_useq;
 
 malloc_mutex_t	prof_dump_mtx;
+static char	prof_dump_prefix[PROF_DUMP_PREFIX_LEN] = {0};
```

Comment:
@interwq When I'm trying to use `base_alloc` in the critical section of `prof_dump_filename_mtx`, the mutex ranks are sort of inverted (base is much lower than prof_dump_filename). What could I do to tackle this?

Reply:
Ah that’s right. We can lock — check if it’s empty — , and if so unlock then alloc with iallocztm, and lock again then try setting it. If some other threads beat us and created it already, unlock and then free the buffer with idalloctm.

Just realized this can only happen in through mallctl, right? This means everything is under the ctl mtx, so we don’t need to worry about allocating the buffer and then deallocate. Try base alloc the buffer (and assert we hold ctl mtx), and assume we don’t need to deallocate should be fine.